### ISSUE TITLE
Comments

### ISSUE THREAD

**Issue by tommoor** [outline/outline#457]
Tracking issue for implementing comments.

**Comment by tommoor** [comment]
### Concept

After researching this week, the current plan is to implement something akin to Etherpad's [atext](http://etherpad.org/doc/v1.6.0/#index_atext) / [apool](http://etherpad.org/doc/v1.6.0/#index_apool) concepts that track additional attributes in the document on a per-character basis by mapping an array of attributes to line and character markers.

Incidentally using a format such as this would also allow us to track contributions by author at a granular level.

### Slate

This additional data would be stored on the document database model, serialized separately from the main Markdown body. This could be implemented as an additional serializer/deserializer pair that handles only attributes, mapping them to/from Slate "Marks".

**Comment by jorilallo** [comment]
@tommoor all this sounds good to me. If someone updates the document via the API without the additional data, should we just clear the metadata/comments at that point? I don't think it will be common, edge case.

**Comment by MarceloCajueiro** [comment]
Hey guys,

How is this going?

We started to use outline at Portabilis (our company) and we are missing this feature. Would be great to have the ability to discuss about a document in order to make it better or just to complement the author if it is something very awaited.

I don't know if we have someone with the necessary skills to help you out in the code but we can help in other ways.

**Comment by almereyda** [comment]
This issue does not seem to be about conversational comments next to a document, but rather about annotations in the text itself.

**Comment by kwix** [comment]
Is this feature still being considered?

**Comment by tommoor** [comment]
It is, it just got delayed behind work on improved permissioning

**Comment by kwstasna** [comment]
@tommoor it would be awesome for an ETA of this feature! As soon as its available our workspace will mosdef move there because you have done an excellent work here guys.

**Comment by tommoor** [comment]
Work has begun, keep me honest 😄

**Comment by pihentagy** [comment]
> Work has begun, keep me honest smile

Commit hash or didn't happen :)

**Comment by milof** [comment]
so is this feature is about disqus type comments below a document (like in the title of this issue) or about google docs like comments in the sidebar (like in the commenting spec https://app.getoutline.com/share/79435f04-c5f3-4a74-8e6f-af5607e5d69d) ???

anyhow I'm really excited about progress on this issue 🤘👍

**Comment by tommoor** [comment]
This feature is for reference based comments, there is no plan to have comments below the document that don't reference the content in some way.

**Comment by milof** [comment]
> This feature is for reference based comments, there is no plan to have comments below the document that don't reference the content in some way.

that's fantastic, thank you

**Comment by milof** [comment]
is this planned to be released alongside collaborative document editing feature?

**Comment by benmort** [comment]
@tommoor amazing news on this upcoming feature. could you do with an assist/peer review/tests around this?

**Comment by almereyda** [comment]
Comments inside the document depend on the data model of the underlying data. Yet good practices have emerged, about how to annotate documents in the web.

There is the [Open Annotation Data Model](http://www.openannotation.org/spec/core/), which allows for interoperability of comments.
And there is a wide annotation ecosystem around [Hypothesis](https://web.hypothes.is/blog/supporting-open-annotation/) that one could consider to integrate with in one way, or the other.

For prior art on the implementation side, we know of two real-time editors that allow comments aside to floating pieces of text. They are:

- CryptPad, on an append-only log; according to the devs it would be rewritten with CRDTs, if ever reimplemented
- Fidus Writer, on a Django application with real-time <> offline support

And there is a semi-synchronised writing environment with comments, that syncs to CouchDB and allows semi-interactive writing sessions, named StackEdit.io.

How do you sketch the data model of document marginalia for Outline in your mind, when thinking of this issue?

**Comment by h4x3rotab** [comment]
How is it going? I think this feature is extremely useful

**Comment by milof** [comment]
been checking this site every couple of weeks
unfortunately looks like the comments feature has been deprioritized

**Comment by Temzzzz1** [comment]
What about comments?

**Comment by almereyda** [comment]
Please also mark this as off-topic:

@pihentagy The screencast says it in the bottom-right: archbee.

**Comment by alex-arabadjiev** [comment]
Has any work been done on this topic? I would love to help out since this is probably the feature we're missing most in Outline. If there's anything the community can do, please let us know!

**Comment by tommoor** [comment]
This feature is dependent on #811 due to the change in data structure allowing storing the comment span metadata. #811 is in alpha, it will probably be about 6 more weeks until it's production ready.

In terms of community involvement I think the main thing is having an understanding of what style of comments people prefer so that implementation can hit the ground running 😄

**Comment by h4x3rotab** [comment]
Inline comments like Google Docs is the best IMO

**Comment by chrisshaw** [comment]
First, I just had a coworker name this as the blocking feature for adoption. Most of his workflows involve collaboratively drafting marketing content and he makes major use of Google Docs' comments.

**DISCLAIMER: I'm part of the Hypothesis team**

Second, given that this is yet to be implemented, I wonder if there's potential for an integration with [Hypothesis](https://github.com/hypothesis)? I hope this comes across as genuine. I've been piloting and pushing for us to adopt Outline as our internal wiki and didn't even make this connection until the aforementioned coworker :point_up: made a joke about us integrating.

H is an open source project (largely FreeBSD). There could be a number of benefits to this approach, including:
  - Users of Outline could combine their internal team notes with snippets they've taken across the web. Think researchers, analysts, journalists...
  - Users' publicly shared Collections could have a built-in forum on the pages themselves
  - Rudimentary support for Outline works out of the box. Obviously there's work to be done here, but still kind of cool: ![image](https://user-images.githubusercontent.com/2966918/141206127-2d4d97f7-2e10-421b-be30-22c66051f441.png)

I'm the new Product Director at H (been here a month) and don't know what I'm doing, so I'm probably speaking out of turn. Oh well. As an Outline user, I love the product (minus some annoying bugs with lists :wink:) and it'd be badass to partner. Let me know if there's appetite, @tommoor, and I'd be happy to connect.

**Comment by aliasgarlabs** [comment]
We at CleverFlow have switched from Notion to Outline and so far we love the simplicity of it. However, we just realized there are no document comments. 

IMHO a single hierarchy comment per document would be simple and sufficient for most use cases, given the nuances of in-line comments. 

@tommoor Let me know how we can help!

**Comment by tommoor** [comment]
<img width="565" alt="image" src="https://user-images.githubusercontent.com/380914/160068682-7447a596-80b0-4d09-a5d0-1fe9996bd481.png">

Very actively being worked on

**Comment by gemathus** [comment]
@tommoor I just checkout this branch and it's looking awesome! I would love to pitch in on this much requested feature. Any tasks you would be willing to allow me to work on?

**Comment by Limezy** [comment]
I've seen it has been tagged on the milestone v0.65. Seems like still a very looooong time to wait 😅

**Comment by tommoor** [comment]
> Any tasks you would be willing to allow me to work on?

@gemathus It's unfortunately difficult to share the work on this one, most of what's left is intricacies of integration with the editor, and design polish. The best way to help is to take on other issues in the backlog so I don't have to 😆

**Comment by gemathus** [comment]
Got it! Good luck!

**Comment by Limezy** [comment]
I was wondering : how will you manage who can write or read comments ? Will the readers be able to view the comments ? Or will comments reading and writing be reserved to the editors of the document ?
In the future having a tracking mode when you could comment not only a word or a paragraph, but also a change you have made would be extremely useful

**Comment by kenske** [comment]
Inline comments are great, but that seems way more complicated than just having a comment section at the bottom of the document. Thoughts on starting with that first?

**Comment by almereyda** [comment]
To comment on the use of Hypothesis for commenting in Outline as outlined by @chrisshaw above, the discussion #3576 now expresses an idea about how to leverage Hypothesis for discussion of specific revisions of *publicly shared* documents.

This would mimic GitHub/-Lab's behaviour of restricting line comments to distinct commit diffs.

<details>

<summary>Easter egg</summary>

Please also find this lovely interview with Chris, which underlines why one may want to integrate and/or interoperate with an open web of annotations:

- https://web.hypothes.is/blog/meet-our-new-director-of-product/

</details>

**Comment by Limezy** [comment]
Oh no 😢 comments are again postponed by 2 months...

**Comment by marijnvandevoorde** [comment]
Shame, I really appreciate Outline. This, and mentioning other users, is what I lack on a daily base. All other things I can totally live with. But once you are working more async and with a bigger team, this becomes a gap. I'm sure it's not a make or break thing for small teams or solo users so I guess it's only addressing a need for some users and therefore has less reach. Still, I'd love to see this :D

Not to be ungrateful, let's be clear that Outline is already really nice.

**Comment by juaneloDev** [comment]
Can anyone share any workarounds they've been using for the lack of comments in the meantime?

**Comment by juaneloDev** [comment]
So, after doing some research, we've identified some third party, open-source commenting implementations that we're considering adding into our Outline deployment in the meantime. I'm linking to that post here: https://github.com/outline/outline/discussions/3878

**Comment by tgy** [comment]
Good luck @tommoor we count on you! 💪🏼

**Comment by edelbende** [comment]
Hi @tommoor,
Any update about this feature ? 🙂

**Comment by NordBeaver** [comment]
Comments are a key feature of any working team. I have now discovered that this is the only thing holding us back from using outline. I understand that work on this feature is on hold? Can we expect it before the end of this year?

**Comment by tommoor** [comment]
We're now testing this feature in the [cloud hosted version of outline](https://app.getoutline.com) – you can enable the comments beta under **Settings -> Features -> Comments.**

**Comment by Limezy** [comment]
@tommoor are you ok if we start filling up issues on that feature, or is it still too soon ?

**Comment by tommoor** [comment]
Yes, okay to file issues

**Comment by crash7** [comment]
@tommoor any estimation when this could be available for the self hosted version? even if it's still in beta.

**Comment by tommoor** [comment]
This feature was released in [v0.69.0](https://github.com/outline/outline/releases/tag/v0.69.0)
